### PR TITLE
Fix testRefreshingMultipleTimesWithinWindowSucceeds

### PR DESCRIPTION
**Pull request by jkakavas**
Previously all the threads were writing the received tokens to a
HashSet. In cases with many threads, sometimes (1 every ~25 tests)
calling `size()` on the HashSet returned 2 even though it seemed to
contain only one String and there was no evidence from logging that
`threadSecurityClient.refreshToken()` ever returned a different
access or refresh token.

This commit changes the test to use a ConcurrentHashMap instead,
checking that we only received one pair of access token/refresh token
eventually. It also adds a check so that we won't take into consideration
tokens that are returned after 30s, hence not in the concurrent refresh
time window.

### CODE REVIEW

**Comment by jkakavas** [comment]
This failed once this morning in
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA=java11,ES_RUNTIME_JAVA=java8fips,nodes=immutable&&linux&&docker/279/console

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 1]
```diff
@@ -425,8 +425,12 @@ public void testRefreshingMultipleTimesWithinWindowSucceeds() throws Exception {
...
                 }
                 threadSecurityClient.refreshToken(refreshRequest, ActionListener.wrap(result -> {
-                    accessTokens.add(result.getTokenString());
-                    refreshTokens.add(result.getRefreshToken());
+                    if (accessTokens.get().size() > 0 && accessTokens.get().contains(result.getTokenString()) == false ||
+                        refreshTokens.get().size() > 0 && refreshTokens.get().contains(result.getRefreshToken()) == false) {
+                        newToken.set(true);
+                    }
+                    accessTokens.get().add(result.getTokenString());
+                    refreshTokens.get().add(result.getRefreshToken());
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java]: The root problem is that `HashSet` is not thread-safe. And wrapping it in an `AtomicReference` won't do any good. I recommend using `ConcurrentHashMap` and keeping the original logic - without `newToken` .

**Review by albertzaharovits — CHANGES_REQUESTED** [review]

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 2]
```diff
@@ -393,8 +392,6 @@ public void testRefreshingMultipleTimesWithinWindowSucceeds() throws Exception {
             UsernamePasswordToken.basicAuthHeaderValue(SecuritySettingsSource.TEST_USER_NAME,
                 SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING)));
         SecurityClient securityClient = new SecurityClient(client);
-        Set<String> refreshTokens = new HashSet<>();
-        Set<String> accessTokens = new HashSet<>();
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java]: `Instant t1 = clock.instant();`

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 3]
```diff
@@ -425,8 +425,12 @@ public void testRefreshingMultipleTimesWithinWindowSucceeds() throws Exception {
                     return;
                 }
                 threadSecurityClient.refreshToken(refreshRequest, ActionListener.wrap(result -> {
-                    accessTokens.add(result.getTokenString());
-                    refreshTokens.add(result.getRefreshToken());
+                    if (accessTokens.get().size() > 0 && accessTokens.get().contains(result.getTokenString()) == false ||
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java]: `Instant t2 = clock.instant();`

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java [hunk 4]
```diff
@@ -425,8 +425,12 @@ public void testRefreshingMultipleTimesWithinWindowSucceeds() throws Exception {
...
+                        refreshTokens.get().size() > 0 && refreshTokens.get().contains(result.getRefreshToken()) == false) {
+                        newToken.set(true);
+                    }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java]: To play it cool, discard responses (don't fail the test, but log something) if `t2 - t1` is greater than 30 sec.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java (hunk 1 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/TokenAuthIntegTests.java]: Gotcha

**Review by albertzaharovits — APPROVED** [review]
LGTM

**Pull request merged**